Create AuxVectorKernel

aeslaughter opened the pull request:
I am going to include documentation for the Aux system in general, but I wanted to start some testing to see if this impacts anything.

(closes #12312)

<!--
INCLUDE THE FOLLOWING IN THE PR DESCRIPTION
- Explain relevant design information for your change.
- Follow the [Coding Standards](http://mooseframework.org/wiki/CodeStandards/).
- Submit or improve [Test Cases](http://mooseframework.org/wiki/MooseTraining/testing/).
- Reference a specific issue, place "refs #<issue>" or "closes #<issue>" (e.g., #closes #1234).
-->

aeslaughter committed:
Create AuxKernelBase to allow for VectorAuxKernel and AuxKernel classes

(refs #12312)

aeslaughter committed:
Create AuxVectorKernel

aeslaughter committed:
WIP: Add documentation for AuxKernel system

aeslaughter committed:
More work on Aux system doco

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/321983/) on 33e32bf wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/13257/site)

This comment will be updated on new commits.

lindsayad reviewed (commented):
Is there a reason why you even need `AuxKernelBase`? Could you just have the template class `AuxKernelTempl` and then just do `typedefs`? E.g.:
```c++
template <ComputeValueType>
class AuxKernelTempl;
typedef AuxKernelTempl<Real> AuxKernel;
typedef AuxKernelTempl<RealVectorValue> AuxVectorKernel;
```
You would have to do some changes anywhere there is a forward declaration of `AuxKernel` but this would cut down our inheritance tree. AFAICT the derived classes that are in this PR now add no functionality beyond the base template.

Also, would `VectorAuxKernel` be more consistent with our vector naming conventions? We have `VectorMooseVariable` for instance. I care less about this

permcody reviewed (commented):
This looks reasonable as long as everyone is ok introducing another layer with templating to one of the most fundamental classes in MOOSE. As I stated there might be another way to implement this with a different named function. I do realize there are drawbacks to that approach because you can't make either of them pure virtual and it might be misleading to have them both "available". 

I'm not saying this design is wrong - just making everyone think critically about the new design.

aeslaughter commented on framework/include/auxkernels/AuxKernelBase.h:
```diff
@@ -0,0 +1,306 @@
+  /**
+   * Compute and return the value of the aux variable.
+   */
+  virtual ComputeValueType computeValue() = 0;
```
It is more than that, the types of _u, _u_old, _u_older are also part of the template, which is actually the main driver for it. If you want _u to work for both versions, then templating is how to get it done.  This follows the same design as regular Kernel objects.

aeslaughter commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -535,40 +667,61 @@ AuxiliarySystem::computeElementalVars(ExecFlagType type)
+  }
 }
+
+template void AuxiliarySystem::computeElementalVarsHelper<AuxKernel>(
```
I just wanted to keep the src in the .C file.

aeslaughter commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -535,40 +667,61 @@ AuxiliarySystem::computeElementalVars(ExecFlagType type)
-{
-  // Evaluate aux variables to get the solution vector
-  compute(EXEC_LINEAR);
+      solution().close();
```
I just followed the design in place, didn't give it any thought.

aeslaughter replied:
It should be possible to combine the vector and non-vector versions, but this would require major changes to the logic in the compute threads. I feel that this is beyond the scope of getting the core functionality in place. But, if you feel strong about it I can take a stab at it.

aeslaughter commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -456,59 +560,87 @@ AuxiliarySystem::computeScalarVars(ExecFlagType type)
+AuxiliarySystem::computeElementalVarsHelper(
+    const MooseObjectWarehouse<AuxKernelType> & warehouse,
+    const std::vector<std::map<std::string, MooseVariableFEBase *>> & vars,
+    const PerfID & timer)
```
I will change these

aeslaughter commented on framework/src/systems/AuxiliarySystem.C:
```diff
@@ -157,13 +182,33 @@ AuxiliarySystem::addVariable(const std::string & var_name,
+    {
+      MooseVariable * var = _vars[tid].getFieldVariable<Real>(var_name);
+
+      if (var != NULL)
```
Sure

aeslaughter commented:
@lindsayad Isn't AuxKernelTempl == AuxKernelBase. I could make the typedefs for AuxKernel and AuxVectorKernel if you perfer, I just made them classes because I prefer to avoid macros even if the code is somewhat useless.

aeslaughter commented:
I chose AuxVectorKernel, because it started with "Aux". I like the consistent that the the versions of aux kernels start with the same prefix. But, I also don't care that much.

aeslaughter commented on framework/include/loops/ComputeNodalAuxBcsThread.h:
```diff
@@ -37,7 +38,9 @@ class ComputeNodalAuxBcsThread
-  const MooseObjectWarehouse<AuxKernel> & _storage;
+  const MooseObjectWarehouse<AuxKernelType> & _storage;
+
+  const std::vector<std::map<std::string, MooseVariableFEBase *>> & _aux_vars;
```
I can update this. I actually think that there is no need for the map at all, just a vector of the pionters would work. It seems like all that is done is the map is iterated over and it.second is called.

lindsayad commented:
> Isn't AuxKernelTempl == AuxKernelBase.

Yes, but if we didn't have have any derived classes then the `Base` name seemed inappropriate.

> I prefer to avoid macros

Not macros, but good old C++ typedefs.

> code is somewhat useless.

Yea, I guess I don't care about this too much either.

aeslaughter committed:
Partially address review comments

- update to use nullptr
- fix reference and const for PerfID objects
- convert variable storage structures to use vector of vectors

aeslaughter commented:
@lindsayad I will change the name to AuxKernelTempl, but I am going to keep the individual classes for AuxKernel and AuxVectorKernel because the AuxKernelTempl has a second argument for the VarFieldKind, this I would have to make a macro rather than use the typedef

lindsayad commented:
In the place where you need to use the `var_type` that you're currently supplying through the constructor parameter, you could do:
```c++
std::is_same<Real, ComputeValueType>::value ? Moose::VarFieldType::VAR_FIELD_STANDARD : Moose::VarFieldType::VAR_FIELD_VECTOR
```

aeslaughter committed:
AuxKernelBase to AuxKernelTempl

aeslaughter commented:
I implemented the is_same, but I still think I want to keep the two classes. Because without them the forward declare for classes deriving from these becomes the following, which for users creating new custom objects isn't very C++ newbie friendly.

```
template <typename ComputeValueType>
class AuxKernelTempl;

typedef AuxKernelTempl<RealVectorValue> AuxVectorKernel;
```

aeslaughter commented:
Although, I guess most people would be creating an AuxKernel/AuxVectorKernel so they would just have '#include "AuxKernel.h' and not need to forward declare.  So, really I don't know what I am doing, I am just going back and forth.

aeslaughter committed:
Utilize std::is_same to clean up need for extra constructor argument

aeslaughter committed:
Use typedefs rather than class definitions for AuxKernel and AuxVectorKernel

aeslaughter committed:
Move all AuxKernel and AuxVectorKernel stuff to AuxKernel.C/h

aeslaughter committed:
AuxVectorKernel -> VectorAuxKernel

aeslaughter committed:
Remove unnecessary if statement

aeslaughter commented:
I implemented all the suggestions, with the exception of trying to consolidate the vec and non-vector compute loops

aeslaughter committed:
Add an input file to show decoupling of equations using AuxKernel/Variable

aeslaughter committed:
Add documentation for AuxKernel system

aeslaughter committed:
Fix removed syntax that is showing up on website

aeslaughter committed:
Update tooltip syntax for materialize 1.0

aeslaughter committed:
Make example and tutorial links optional

aeslaughter committed:
Update gold for MooseDocs bug fixes

aeslaughter committed:
Add requirements to test spec of step 6
